iot_config/src/gateway_service.rs

```diff
@@ -1,17 +1,130 @@
-use crate::GrpcResult;
-use helium_proto::services::iot_config::{
-    self, GatewayRegionParamsReqV1, GatewayRegionParamsResV1,
+use crate::{
+    region_map::{self, RegionMap},
+    GrpcResult, Settings,
 };
-use tonic::Request;
+use anyhow::Result;
+use file_store::traits::MsgVerify;
+use helium_crypto::{Keypair, PublicKey, Sign};
+use helium_proto::{
+    services::iot_config::{
+        self, GatewayRegionParamsReqV1, GatewayRegionParamsResV1, LoadRegionReqV1, LoadRegionResV1,
+    },
+    Message, Region,
+};
+use hextree::Cell;
+use node_follower::{
+    follower_service::FollowerService,
+    gateway_resp::{GatewayInfo, GatewayInfoResolver},
+};
+use sqlx::{Pool, Postgres};
+use tonic::{Request, Response, Status};
+
+pub struct GatewayService {
+    admin_pubkey: PublicKey,
+    follower_service: FollowerService,
+    pool: Pool<Postgres>,
+    region_map: RegionMap,
+    signing_key: Keypair,
+}
+
+impl GatewayService {
+    pub async fn new(settings: &Settings) -> Result<Self> {
+        let pool = settings.database.connect(10).await?;
+        let region_map = RegionMap::new(&pool).await?;
+        Ok(Self {
+            admin_pubkey: settings.admin_pubkey()?,
+            follower_service: FollowerService::from_settings(&settings.follower),
+            pool,
+            region_map,
+            signing_key: settings.signing_keypair()?,
+        })
+    }
 
-pub struct GatewayService {}
+    fn verify_admin_signature<R>(&self, request: R) -> Result<R, Status>
+    where
+        R: MsgVerify,
+    {
+        request
+            .verify(&self.admin_pubkey)
+            .map_err(|_| Status::permission_denied("invalid admin signature"))?;
+        Ok(request)
+    }
+}
 
 #[tonic::async_trait]
 impl iot_config::Gateway for GatewayService {
     async fn region_params(
         &self,
-        _request: Request<GatewayRegionParamsReqV1>,
+        request: Request<GatewayRegionParamsReqV1>,
     ) -> GrpcResult<GatewayRegionParamsResV1> {
-        unimplemented!()
+        let request = request.into_inner();
+
+        let pubkey = PublicKey::try_from(request.address.clone())
+            .map_err(|_| Status::invalid_argument("invalid gateway address"))?;
+        request
+            .verify(&pubkey)
+            .map_err(|_| Status::permission_denied("invalid request signature"))?;
+
+        let GatewayInfo { location, gain, .. } = self
```

Comment:
I assume the use of the follower here in temporary, until a helius db instance is available ?

Reply:
yeah this allows gateway-rs to use the config service and remove validators from the equation until switchover time.